src/Components/Server/src/Circuits/ComponentParameterDeserializer.cs

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Text.Json;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Components.Server
+{
+    internal class ComponentParameterDeserializer
+    {
+        private readonly ILogger<ComponentParameterDeserializer> _logger;
+        private readonly ComponentParametersTypeCache _parametersCache;
+
+        public ComponentParameterDeserializer(
+            ILogger<ComponentParameterDeserializer> logger,
+            ComponentParametersTypeCache parametersCache)
+        {
+            _logger = logger;
+            _parametersCache = parametersCache;
+        }
+
+        public bool TryDeserializeParameters(IList<ComponentParameter> parametersDefinitions, IList<object> parameterValues, out ParameterView parameters)
+        {
+            parameters = default;
+            var parametersDictionary = new Dictionary<string, object>();
+
+            if (parameterValues.Count != parametersDefinitions.Count)
+            {
+                // Mismatched number of definition/parameter values.
+                Log.MismatchedParameterAndDefinitions(_logger, parametersDefinitions.Count, parameterValues.Count);
+                return false;
+            }
+
+            for (var i = 0; i < parametersDefinitions.Count; i++)
+            {
+                var definition = parametersDefinitions[i];
+                if (definition.Name == null)
+                {
+                    Log.MissingParameterDefinitionName(_logger);
+                    return false;
+                }
+
+                if (definition.TypeName == null && definition.Assembly == null)
```

Comment:
It seems unnecessary that we should shuttle the type and assembly names of the parameters back and forth. We have to be able to find the component type on the server - we should use that instead of transferring more data.

Reply:
Longer discussion about this below, so marking this one as resolved to avoid forking the conversation.